### PR TITLE
feat: add local sign/list/verification for OCI layout directory

### LINKED ISSUE #283: notation sign/verify to work with local storage
Implement notation to work with signatures in the local storage/filesystem.

### PR DESCRIPTION
**Pull request by patrickzheng200**
This PR adds local sign/list/verification for OCI image layout directory. 
For RC.4:
1. It only supports storing the generated signature into the target OCI layout directory.
2. It supports listing signatures within the OCI layout directory.
3. It only supports verifying signatures within the target OCI layout directory. 

This PR is based on spec PR: https://github.com/notaryproject/notation/pull/601 (Merged).

This PR is dependent on the corresponding notation-go PR: https://github.com/notaryproject/notation-go/pull/288. 
Please review the notation-go PR first.

Resolves #283.

Both remote registry and oci-layout scenario are tested. E2E tests are also included.

### CODE REVIEW

**Review by qweeah — CHANGES_REQUESTED** [review]

File: cmd/notation/sign.go
Comment (qweeah) [inline cmd/notation/sign.go:195]: Looks like this code block is high similar to `prepareRemoteSigningContent`, can we decouple the difference between registry and OCI layout out and use a single prepare function? Most of the difference is on subject reference parsing.

**Review by byronchien — DISMISSED** [review]

File: cmd/notation/sign.go
Comment (patrickzheng200) [reply cmd/notation/sign.go:67]: IMO, for users who are not familiar with OCI layout, the original description points a way to the position of the target manifest descriptor, i.e., the entry point of the whole OCI layout structure. Will remove `local` as suggested.

File: cmd/notation/sign.go
Comment (patrickzheng200) [reply cmd/notation/sign.go:70]: That's a good point. Will update the code.

File: cmd/notation/list.go
Comment (shizhMSFT) [inline cmd/notation/list.go:83]: This code block can be refactored as they have many common parts.

File: cmd/notation/list.go
Comment (shizhMSFT) [inline cmd/notation/list.go:65]: Note: In general, `Resolve()` takes a tag or a digest and should not take a full reference while `Repository.Resolve()` can take any form.

File: cmd/notation/sign.go
Comment (shizhMSFT) [inline cmd/notation/sign.go:115]: `signLocal` and `signRemote` should be common and should only have slight difference. Thus refactoring is requested.

File: cmd/notation/verify.go
Comment (shizhMSFT) [inline cmd/notation/verify.go:116]: same here

File: cmd/notation/registry.go
Comment (patrickzheng200) [reply cmd/notation/registry.go:259]: @shizhMSFT The logic here is getting the `last index` of the separator in the string. I'm not sure if `strings.Cut` is able to do it, as it seems to cut around the `first instance` of sep?

Comment (shizhMSFT) [reply cmd/notation/registry.go:259]: That makes sense.

File: cmd/notation/list.go
Comment (patrickzheng200) [reply cmd/notation/list.go]: Updated to `--oci-layout` flag based on our discussion.

File: go.mod
Comment (patrickzheng200) [reply go.mod:32]: Yes, this line will be removed once that PR is merged.

File: cmd/notation/manifest.go
Comment (patrickzheng200) [reply cmd/notation/manifest.go:90]: Correct. Running `localArtifactReference("hello", "sha256:abcde")` or `localArtifactReference("./hello", "sha256:abcde")` would print out `./hello@sha256:abcde` as the result.
This is a helper function to generate print out for commands such as `notation list`. It is also used as the `ArtifactReference` field in `SignOptions` in the OCI layout case to pass the `ParseReference` check in notation-go `Sign`.
I will rename the `reg` and `repo`, as they might look confusing for OCI layout.

File: cmd/notation/verify.go
Comment (patrickzheng200) [reply cmd/notation/verify.go:130]: Renamed to as suggested.

File: cmd/notation/manifest.go
Comment (patrickzheng200) [reply cmd/notation/manifest.go:90]: Yeah, I found a bug around this. Updating the code.

Comment (patrickzheng200) [reply cmd/notation/manifest.go:90]: Removed the function.

File: cmd/notation/manifest.go
Comment (JeyJeyGao) [inline cmd/notation/manifest.go:93]: (i == 1 && len(raw) > 2 && unicode.IsLetter(rune(raw[0])) && raw[2] == '\\')  why we need this?

File: cmd/notation/manifest.go
Comment (JeyJeyGao) [inline cmd/notation/manifest.go:88]: Can we keep the same code stype for line 96 & line 88

File: cmd/notation/manifest.go
Comment (JeyJeyGao) [inline cmd/notation/manifest.go:98]: Is it possible that `ref` is empty?

File: internal/testdata/oci-layout/blobs/sha256/63b65145d645c1250c391b2d16ebe53b3747c295ca8ba2fcb6b0cf064a4dc21c
Comment (patrickzheng200) [reply internal/testdata/oci-layout/blobs/sha256/63b65145d645c1250c391b2d16ebe53b3747c295ca8ba2fcb6b0cf064a4dc21c:1]: We are not using this testdata anymore. Removed it.

File: cmd/notation/manifest.go
Comment (patrickzheng200) [reply cmd/notation/manifest.go:93]: This is for Windows path starting with a disk name such as `D:\oci-layout`.

File: cmd/notation/manifest.go
Comment (patrickzheng200) [reply cmd/notation/manifest.go:98]: Actually, `if path == ""` should also be moved out of the `else` statement.

File: cmd/notation/manifest.go
Comment (shizhMSFT) [inline cmd/notation/manifest.go:56]: I just noticed `ArtifactReference` is never used in `Signer.Sign()` but in `notation.Sign()`. Can you recall why we have `ArtifactReference` in `SignerSignOptions` in the first place?

Comment (shizhMSFT) [reply cmd/notation/manifest.go:56]: Also, `ArtifactReference` does not need to be a full reference for `Sign()`.

Comment (shizhMSFT) [reply cmd/notation/manifest.go:56]: Do you think we need to update `notation-go` accordingly?

File: cmd/notation/manifest.go
Comment (patrickzheng200) [reply cmd/notation/manifest.go:25]: Good idea. Will do.

File: cmd/notation/manifest.go
Comment (shizhMSFT) [reply cmd/notation/manifest.go:56]: SGTM

**Review by shizhMSFT — COMMENTED** [review]
LGTM but pending https://github.com/notaryproject/notation-go/pull/295

**Comment by yizha1** [comment]
@patrickzheng200 @qweeah, will this flag `--oci-layout` and `--scope` be marked as experimental in this PR?

**Comment by patrickzheng200** [comment]
> @patrickzheng200 @qweeah, will these flags `--oci-layout` and `--scope` be marked as experimental in this PR?

@yizha1 Yes, that's the plan. This PR should be merged after https://github.com/notaryproject/notation/pull/613.

File: go.mod
Comment (patrickzheng200) [reply go.mod]: Resolving because https://github.com/notaryproject/notation-go/pull/288 has been merged. Updated dependency of this PR.

File: cmd/notation/manifest.go
Comment (patrickzheng200) [reply cmd/notation/manifest.go:79]: Yes. It's an unexported helper function, but yeah, renamed it to `resolveArtifactDigestReference`

File: cmd/notation/manifest.go
Comment (patrickzheng200) [reply cmd/notation/manifest.go]: @priteshbandi I tried this method, it seems not working on cases like `localhost:5000/myRepo@sha256:xxx`. In this example, `ref.Registry` is `localhost:5000` and `ref.Repository` is `myRepo`. However, the output of `url.JoinPath(ref.Registry, ref.Repository)` is just `localhost:5000`.  
(The reason is that, when doing url.JoinPath, one of the steps is parsing `localhost:5000`. Due to the existence of `:`, `localhost` will be parsed as the url scheme.)

File: cmd/notation/sign.go
Comment (shizhMSFT) [inline cmd/notation/sign.go:148]:
```suggestion
				fmt.Fprintln(os.Stderr, "Warning: Removal of outdated referrers index from remote registry failed. Garbage collection may be required.")
```

Comment (shizhMSFT) [reply cmd/notation/sign.go:148]: The error can be anything, not just 405 method not allowed.

**Review by priteshbandi — APPROVED** [review]
lgtm

File: cmd/notation/sign.go
Comment (patrickzheng200) [reply cmd/notation/sign.go]: Updated the code.

**Review by priteshbandi — APPROVED** [review]

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**
